### PR TITLE
define SYS_getrandom if not available in Makefile

### LINKED ISSUE #72: compatibility with Asuswrt-Merlin
hi,

I was trying to compile dsvpn on my router (Asuswrt-Merlin, 384.14_2) and it does not contain any linux headers and `SYS_getrandom` is not defined in headers of GCC either which is installed by [`opkg`](https://github.com/Entware/Entware) (package manager on Asuswrt-Merlin). So I need to define it manually:

```
opkg install gcc
make CC=gcc OPTFLAGS="-DSYS_getrandom=278"
```

Maybe we can define it if the arch is arm64, like this:

```
+++ b/src/charm.c
@@ -328,6 +328,9 @@ void uc_memzero(void *buf, size_t len)
 void uc_randombytes_buf(void *buf, size_t len)
 {
 #ifdef __linux__
+#if defined(__aarch64__) && !defined(SYS_getrandom)
+#define SYS_getrandom 278
+#endif
     if ((size_t) syscall(SYS_getrandom, buf, (int) len, 0) != len) {
         abort();
     }

```

What do you think?

### PR DESCRIPTION
**Pull request by cofyc**
fixes #72 
Signed-off-by: Yecheng Fu <fuyecheng@pingcap.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -27,6 +27,14 @@ $(CFLAGS_FILE):
 				$(CC) $${CFLAGS} $${flag} "$(COMPILE_TEST_FILE)" >/dev/null 2>&1 && CFLAGS="$$CFLAGS $$flag"; \
 			done; \
 			CFLAGS="$${CFLAGS} -Wall -W -Wshadow -Wmissing-prototypes"; \
+			echo "#include <unistd.h>\n#include <sys/syscall.h>\nint main(void) { char buf[32] = {0}; syscall(SYS_getrandom, buf, 32, 0); }" > "$(COMPILE_TEST_FILE)"; \
+			if ! $(CC) "$(COMPILE_TEST_FILE)" >/dev/null 2>&1; then \
+				arch=$$(uname -m); \
+				case $$arch in \
+	              aarch64) CFLAGS="$$CFLAGS -DSYS_getrandom=278";; \
+				  x86_64) CFLAGS="$$CFLAGS -DSYS_getrandom=318";; \
```
Comment (cofyc) [inline Makefile]: I'm not sure the exact output of `uname -m` in other architectures. We can support more if there are other systems need this hack.

**Comment by llelf** [comment]
What if it’s missing because…the syscall is just not there (old kernel)?

**Comment by cofyc** [comment]
> What if it’s missing because…the syscall is just not there (old kernel)?

the `syscall(SYS_getrandom, ...)` call will fail. (`-1` is returned).

If we want `dsvpn` to work on old kernels, we need to fall back to other random methods.

**Comment by cofyc** [comment]
I'm going to close this. As it's better if the code itself keeps assuming a modern and sane environment as suggested by @jedisct1 [here](https://github.com/jedisct1/dsvpn/issues/72#issuecomment-581027545).

**Pull request closed without merging**